BloomFilter.md

```diff
@@ -103,18 +123,22 @@ To calculate the size the filter should be for another false positive rate `p`,
 following formula. The output is in bits per distinct element:
 
 ```c
--8 / log(1 - pow(p, 1.0 / 8));
+c = -8 / log(1 - pow(p, 1.0 / 8))
 ```
 
+In the real scenario, the size of the Bloom filter and the false positive rate may vary from
+different implementations. It is recommended to set false positive to 1% so that a Bloom filter
+with 1.2MB size can contain one million distinct values, which should satisfy most cases according
+to default row group size. It is also recommended to expose the ability for setting these
+parameters to end users.
+
 #### File Format
-The Bloom filter data of a column is stored at the beginning of its column chunk in the
-row group. The column chunk metadata contains the Bloom filter offset. The Bloom filter is
-stored with a header containing the size of the filter in bytes, the algorithm, and the
-hash function.
-
-### Reference
-1. [Bloom filter introduction at Wiki](https://en.wikipedia.org/wiki/Bloom_filter)
-2. [Cache-, Hash- and Space-Efficient Bloom Filters](http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf)
-3. [A Reliable Randomized Algorithm for the Closest-Pair Problem](http://www.diku.dk/~jyrki/Paper/CP-11.4.1997.ps)
-4. [xxHash](https://cyan4973.github.io/xxHash/)
-5. [Network Applications of Bloom Filters: A Survey](https://www.eecs.harvard.edu/~michaelm/postscripts/im2005b.pdf)
+The Bloom filter data of a column, which contains the size of the filter in bytes, the algorithm,
+the hash function and the Bloom filter bitset, is stored near the footer. The Bloom filter data
+offset is stored in column chunk metadata.
+
+#### Encryption
+The Bloom filter offset is stored in column chunk metadata which will be encrypted with the column
+key when encryption is enabled. The Bloom filter data itself should also be encrypted with column
```

Comment:
Bloom filters are stored similarly to pages - with a page header (of the BloomFilterPageHeader type), and a page "data", or bitset. Therefore, they will be encrypted in the same way as we encrypt pages today - encrypting the PageHeader structure, and encrypting the page (bitset) itself, both with the same column key, but with different AAD module types - "BloomFilter PageHeader" (8) and "BloomFilter Page" (9) (let me know if you prefer the latter to be called "BloomFilter Bitset") . I'll create a pull request to update the encryption spec with these 2 new types.

Also, it would be good to see the BloomFilter Thrift structures listed and explained in this doc (Bloom filter spec), in the previous section (File Format). It will be useful both in general, and in particular for explanation of their encryption.

Reply:
I 'm OK with BloomFilter Page, it looks pairwise. Will add thrift in next commit.

Reply:
Sounds good. By the way, a question - what are the typical and maximal sizes of the bitsets? (is 
 there is a way to estimate these?) If not too big, we might always encrypt them with GCM cipher (to make them tamper-proof), in both encryption algorithms.

Reply:
If we set the false positive rate to 1%,  1.2MB bitset could hold 1 million distinct values. So I would take 1MB as the typical maximal size of Bloom filter bitset, that should satisfy most of the cases.

Reply:
Cool. If this is the order of magnitude of max size of chunk's  bitset (and since, by design, Bloom filters are smaller than dictionary pages), we don't have to treat Bloom filter encryption identically to pages, and can apply maximal protection even in the relaxed GCM_CTR algorithm. 

More specifically, looks like Bloom filters have two serializable modules (if I got this right). 

- One is the PageHeader thrift structure (with its internal fields, including `optional BloomFilterPageHeader bloom_filter_page_header`). This structure is serialized by Thrift, and written to file output stream, somewhere close to the footer. 

- It is followed by the Bitset, serialized and written right after the filter header.

For filters in sensitive columns, both the PageHeader structure and the Bitset will be encrypted 
 using the AES GCM cipher, with the same column key, but with different AAD module types - "BloomFilter PageHeader" (8) and "BloomFilter Bitset" (9).

Reply:
Even better, in encryption we can use a module name "BloomFilter Header" instead of "BloomFilter PageHeader", to remove any association with page encryption.

Reply:
Here's a calculation [excel ](https://docs.google.com/spreadsheets/d/1LQqGZ1EQSkPBXtdi9nyANiQOhwNFwqiiFe8Sazclf5Y/edit#gid=2008252305)created by Ryan, you may find more info from there.

It worth to note that the recommended default maximal size of the Bloom filter may not cover extreme case, such as a table with one column of int32 type, which may have more than 32 million distinct values. So we open the configuration to the users to choose the maximal size by themselves.

I'm OK with using "BloomFilter Header". Thanks for working on this:)

Reply:
Welcome :) 
The excel stops at 2MB, which is much smaller than a total of chunk data pages. When users configure the max size, their goal will also be to make filter much smaller than the data; otherwise its just easier to read all data pages. Moreover, "using less space than dictionaries" is the formal goal of this spec. So I think we should be fine with second proposal.

Reply:
The second proposal can be written up along these lines -  
_Bloom filters have two serializable modules - the PageHeader thrift structure (with its internal fields, including the `BloomFilterPageHeader bloom_filter_page_header`), and the Bitset. The header structure is serialized by Thrift, and written to file output stream; it is followed by the serialized Bitset._ 

_For Bloom filters in sensitive columns, each of the two modules will be encrypted after serialization, and then written to the file. The encryption will be performed using the AES GCM cipher, with the same column key, but with different AAD module types - "BloomFilter Header" (8) and "BloomFilter Bitset" (9). The length of the encrypted buffer is written before the buffer, as described in the Parquet encryption specification._

So, to sum this up, you have two proposals, with the following trade-off:
somewhat simpler design -vs - protection against bitset tampering attacks in GCM_CTR encryption algorithm.

The performance will be identical in all practical usecases. In extreme cases where bitset size is comparable to data size, the throughput of the first proposal would be somewhat higher (only in GCM_CTR algo, and only in Java 8 and below) - but then, I guess, encryption throughput would be the least of your worries, you dont want users to get to this extreme.

Reply:
Thanks, @ggershinsky, Nice writeup.

As your summary in last, I would prefer to adopt AES GCM as well. So you will update Encryption.md to include these, right?

Reply:
Yep, I'll open a jira/pr to update Encryption.md with the new module types for Bloom filter encryption.

Reply:
https://github.com/apache/parquet-format/pull/141